fea-rs/src/compile/tables/var_store.rs

```diff
@@ -0,0 +1,312 @@
+//! Building the ItemVariationStore
+//!
+//! this will probably get moved to write-fonts at some point?
+
+use std::collections::HashMap;
+
+use indexmap::IndexMap;
+use write_fonts::tables::variations::{
+    ItemVariationData, ItemVariationStore, VariationRegion, VariationRegionList,
+};
+
+///
+/// this probably ends up living in write-fonts
+#[derive(Clone, Debug, Default)]
+pub(crate) struct VariationStoreBuilder {
+    // region -> index map
+    all_regions: HashMap<VariationRegion, usize>,
+    // we use an index map so that we have a determenistic ordering,
+    // which lets us write better tests
+    delta_sets: IndexMap<RegionSet, IndexMap<DeltaSet, DeltaKey>>,
+    next_id: DeltaKey,
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) struct VariationStoreKeyLookup {
+    map: HashMap<DeltaKey, DeltaKey>,
+}
+
+/// a subset of regions, represented as indicies into the region map
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+struct RegionSet {
+    indicies: Vec<u16>,
+}
+
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+struct DeltaSet {
+    values: Vec<i32>,
+}
+
+/// A key that identifies a particular deltaset in an ItemVariationStore.
+///
+/// This corresponds to the outer/inner mapping used to identify deltas.
+/// While constructing a variation store we assign a temporary key to inserted
+/// deltas; these are reordered when the store is built, at which point any
+/// items that reference the store will need to be remapped from their temporary
+/// key to their final one.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
+pub(crate) struct DeltaKey {
+    pub(crate) outer: u16,
+    pub(crate) inner: u16,
+}
+
+impl DeltaKey {
+    #[allow(dead_code)]
+    pub const NO_DELTAS: DeltaKey = DeltaKey {
+        outer: 0xFFFF,
+        inner: 0xFFFF,
+    };
+
+    const fn next(self) -> Self {
+        let DeltaKey { outer, inner } = self;
+        let (outer, inner) = if outer < 0xFFFF {
+            (outer + 1, inner)
+        } else {
+            (outer, inner + 1)
```

Comment:
um actually this whole function is busted and I'll rewrite it, it's currently only capable of representing `0xFFFF + 0xFFFF` values, instead of using all 32 bits 😬